components/support/jwcrypto/src/direct.rs

```diff
@@ -0,0 +1,227 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Support for "Direct Encryption with a Shared Symmetric Key"
+//! See https://tools.ietf.org/html/rfc7518#section-4.5 for all the details.
+
+use crate::{
+    aes,
+    error::{JwCryptoError, Result},
+    Algorithm, CompactJwe, EncryptionAlgorithm, JweHeader, Jwk, JwkKeyParameters,
+};
+use rc_crypto::rand;
+
+impl Jwk {
+    /// Create a new random key suitable for `Direct` symmetric encryption.
+    /// Consumers can store this for later use, probably via serde serialization.
+    pub fn new_direct_key(kid: Option<String>) -> Result<Self> {
+        // We only support AES256 which has a 32byte key.
+        let mut bytes: Vec<u8> = vec![0; 32];
+        rand::fill(&mut bytes)?;
+        Ok(Jwk {
+            kid,
+            key_parameters: JwkKeyParameters::Direct {
+                k: base64::encode_config(&bytes, base64::URL_SAFE_NO_PAD),
+            },
+        })
+    }
+
+    // Create a new Jwk given the raw bytes of a `Direct` key. We generally
+    // prefer consumers to use serde with the entire key and the only use-case
+    // is currently tests, so let's limit this to tests for now...
+    #[cfg(test)]
+    pub fn new_direct_from_bytes(kid: Option<String>, key: &[u8]) -> Self {
+        Jwk {
+            kid,
+            key_parameters: JwkKeyParameters::Direct {
+                k: base64::encode_config(&key, base64::URL_SAFE_NO_PAD),
+            },
+        }
+    }
+}
+
+pub(crate) fn encrypt_to_jwe(
+    data: &[u8],
+    enc: EncryptionAlgorithm,
```

Comment:
Nit: why accept `enc` as an argument here? If we get any other value than `A256GCM` then we're going to create an invalid JWE, so it seems like we may as well just hard-code it to that unless and until we implement other `enc` schemes in future. (Alternately, we could assert that it has the expected value, to give us an obvious place to add support for other schemes in future).

Reply:
Concretely, if we're going to accept it as an argument, then we should have something like this to dispatch based on it at the end of the function:

```
match enc {
  EncryptionAlgorithm::A256GCM => aes::aes_gcm_encrypt(data, protected_header, &secret),
  // Other encryption algorithms would be added here in future.
}
```